ols/src/llms/utils.py

```diff
@@ -0,0 +1,42 @@
+from langchain.callbacks.base import BaseCallbackHandler
+from langchain_core.outputs.llm_result import LLMResult
+from typing import Any, Dict, List
+import logging
+
+logger = logging.getLogger(__name__)
+
+class GenericTokenCounter(BaseCallbackHandler):
+    def __init__(self, llm):
+        self.llm = llm
+        self.input_tokens = 0
+        self.output_tokens = 0
+
+    def on_llm_start(
+        self, serialized: Dict[str, Any], prompts: List[str], **kwargs: Any
+    ):
+        for p in prompts:
+            self.input_tokens += self.llm.get_num_tokens(p)
+
+    def on_llm_end(self, response: LLMResult, **kwargs: Any):
+        results = response.flatten()
+        input_tokens_count = 0
+        for r in results:
+            if "token_usage" in r.llm_output:
+                token_usage = r.llm_output["token_usage"]
+                # typical token_usage: {'prompt_tokens': 252, 'completion_tokens': 4, 'total': 256}
+                if 'prompt_tokens' in token_usage:
+                    input_tokens_count += token_usage['prompt_tokens']
+
+                if  'completion_tokens' in token_usage:
+                    self.output_tokens += token_usage['completion_tokens']
+                else:
+                    self.output_tokens += self.llm.get_num_tokens(r.generations[0][0].text)
+
+                continue
```

Comment:
interesting nested blocks, will be a bit tricky to test it

Reply:
Maybe just
```python
....
else:
   # fallback to counting
   self.output_tokens += self.llm.get_num_tokens(r.generations[0][0].text)
```
instead of `continue` would be a bit more readable.

Reply:
Let me add a unit test about these callbacks in new commit. 
I tend to use less indentation cause I like put editors side by side on my 4:3 screen ;) I will use `else` block in the update.